jumanji/environments/logic/graph_coloring/viewer.py

```diff
@@ -0,0 +1,223 @@
+
+    def _compute_attractive_forces(
+        self,
+        graph: chex.Array,
+        attractive_forces: np.ndarray,
+        pos: np.ndarray,
+        k: float,
+    ) -> np.ndarray:
+        for i in range(self._num_nodes):
+            for j in range(self._num_nodes):
+                if graph[i, j]:
+                    delta = pos[i] - pos[j]
+                    distance = np.linalg.norm(delta)
+                    direction = delta / (distance + 1e-6)
+                    force = distance * distance / k
+                    attractive_forces[i] -= direction * force
+                    attractive_forces[j] += direction * force
+
+        return attractive_forces
+
+    def _spring_layout(
+        self, graph: chex.Array, seed: int = 42
+    ) -> List[Tuple[float, float]]:
+        """
+        Compute a 2D spring layout for the given graph using
+        the Fruchterman-Reingold force-directed algorithm.
+
+        The algorithm computes a layout by simulating the graph as a physical system,
+        where nodes are repelling each other and edges are attracting connected nodes.
+        The method minimizes the energy of the system over several iterations.
+
+        Args:
+            graph: A Graph object representing the adjacency matrix of the graph.
+            seed: An integer used to seed the random number generator for reproducibility.
+
+        Returns:
+            A list of tuples representing the 2D positions of nodes in the graph.
+        """
+        rng = np.random.default_rng(seed)
+        pos = rng.random((self._num_nodes, 2)) * 2 - 1
+
+        iterations = 100
+        k = np.sqrt(1 / self._num_nodes)
+        temperature = 2.0  # Added a temperature variable
+
+        for _ in range(iterations):
+            repulsive_forces = self._compute_repulsive_forces(
+                np.zeros((self._num_nodes, 2)), pos, k
+            )
+            attractive_forces = self._compute_attractive_forces(
+                graph, np.zeros((self._num_nodes, 2)), pos, k
+            )
+
+            pos += (repulsive_forces + attractive_forces) * temperature
+            # Reduce the temperature (cooling factor) to refine the layout.
+            temperature *= 0.9
+
+            pos = np.clip(pos, -1, 1)  # Keep positions within the [-1, 1] range
+
+        return [(float(p[0]), float(p[1])) for p in pos]
+
+    def _get_fig_ax(self, ax: Optional[plt.Axes]) -> Tuple[plt.Figure, plt.Axes]:
+        if ax is None:
+            fig, ax = plt.subplots(figsize=(self.node_scale, self.node_scale))
+            plt.title(f"{self._name}")
+        else:
+            fig = ax.figure
+            ax.clear()
+        return fig, ax
+
+    def _render_nodes(self, ax: plt.Axes, state: State) -> None:
+        # Set the radius of the nodes as a fraction of the scale,
+        # so nodes appear smaller when there are more of them.
+        node_radius = 0.05 * 5 / self.node_scale
+
+        pos = self._spring_layout(state.adj_matrix)
```

Comment:
Sorry it is my fault for misleading you, by computing the `pos` here and in `_render_edges` the layout has to be computed twice. Computation is not very critical for the render, but I am more worried by the fact that the spring_layout is stochastic, and I don't think it is very robust to rely on the seed to ensure that `pos` is the same in `_render_nodes` and `_render_edges`.

Maybe `pos` can be an argument to both functions ? or stored in `self.pos` ?

Reply:
yeah both work. @clement-bonnet any suggestions on this?

Reply:
Agree! Having `pos` as an argument to both functions looks the best to me.